File: src/main/resources/static/custom-element.js
class CustomElement extends HTMLElement {
  constructor() {
    super();
    this.attachShadow({ mode: 'open' });
    this.shadowRoot.innerHTML = `
      <style>
        :host {
          display: block;
          padding: 10px;
          border: 1px solid #ccc;
        }
      </style>
      <div id="content">initial content</div>
    `;
  }

  setText(text) {
    const contentDiv = this.shadowRoot.getElementById('content');
    contentDiv.textContent = text;
  }

  getText() {
    const contentDiv = this.shadowRoot.getElementById('content');
    return contentDiv.textContent;
  }
}

customElements.define('custom-element', CustomElement);

File: src/main/java/org/webforj/component/CustomElement.java
package org.webforj.component;

import com.webforj.annotation.JavaScript;
import com.webforj.component.element.ElementComposite;
import com.webforj.component.element.annotation.NodeName;

@NodeName("custom-element")
@JavaScript("ws://custom-element.js")
public class CustomElement extends ElementComposite {
  
  public void setText(String text) {
    getBoundComponent().callJsFunction("setText", text);
  }

  public Object getText() {
    return getBoundComponent().callJsFunction("getText");
  }
}
